Allow demographic models to store mutation rates

apragsdale opened the pull request:
See #557.

The issue: implemented demographic models were originally inferred with certain assumptions about mutation rates, generation times, etc, which then are used to calibrate population size and demographic event timings. Since #556, #631, users can specify and simulate with contigs with user-defined mutation rates, so that mutation rates could be set to match what was assumed when a model was first inferred. However, unless a user is very familiar with the literature, they might not be aware that the species' default mutation rate does not match the model they want to use, so that outputs will have measures of diversity that are far too large or too small for that species without realizing it.

One solution, as proposed in #557, is to allow models to have their own mutation rates that can be stored and used instead of the default rate. The default behavior of this PR is to use the species' default rate if a mutation rate is not specified, but now mutation rates would be attributes of the demographic model that can be easily passed to `species.get_contig()`.

For example:
```python
import stdpopsim
species = stdpopsim.get_species("HomSap")
model = species.get_demographic_model("OutOfAfrica_3G09")
contig = species.get_contig("chr1", mutation_rate=model.mutation_rate)
```

This contig now has a mutation rate of 2.35e-8 instead of 1.29e-8. There are some models that don't have associated mutation rates (for example, they were calibrated using recombination rates), and those models can be left without mutation rates so that they use the species' default rate (`None` would default to species' default). This would require going back through and adding mutation rates to each species' demographic model and then have them be QC'd and verified, but I think it would be worth the effort.

I didn't make a ton of changes to each demographic model here or implement tests, since feedback would be really helpful here before spending the time to do that.

grahamgower commented:
Looks like a sensible approach. Thanks @apragsdale!

jeromekelleher commented:
Yep, LGTM also. Just needs a few tests.

petrelharp commented:
Hm - it'd be nice if a mutation rate in the model would be used directly instead of having to remember to pass it in. (I agree, it needs to be an attribute of the model, though, so this is good.) Maybe simulation should throw a warning if the model's mutation rate and the contig's mutation rate don't match?

grahamgower commented:
> Hm - it'd be nice if a mutation rate in the model would be used directly instead of having to remember to pass it in.

That's certainly an option. We'd want to make sure the citation for the species mutation rate isn't printed too.

apragsdale commented:
> Hm - it'd be nice if a mutation rate in the model would be used directly instead of having to remember to pass it in. (I agree, it needs to be an attribute of the model, though, so this is good.) Maybe simulation should throw a warning if the model's mutation rate and the contig's mutation rate don't match?

> That's certainly an option. We'd want to make sure the citation for the species mutation rate isn't printed too.

Those are good points. Agree that using the model parameter as default would be good, but I'm not sure how to use it directly without completely pulling the mutation rate out of the contig class and moving it to the model entirely. We still want to preserve users' abilities to set custom mutation rates as well. A warning makes sense and would avoid a giant refactor

apragsdale commented:
I'm going to go through and add mutation rates to the catalog models here. They will need to be QC'd by someone else most likely, and mutation rate equality between the models and their QC counterparts should also be checked in the tests.

petrelharp commented on stdpopsim/engines.py:
```diff
@@ -112,6 +112,15 @@ def _warn_zigzag(self, demographic_model):
+                "The demographic model mutation rate differs from the contig mutation "
+                "rate (simulation uses the contig mutation rate). Diversity levels "
+                "may be different than expected for this species. For details see: "
+                "(url of some documentation for this)"
```
maybe stick in the mutation rates? Like `f"The demographic model has mutation rate {demographic_model.mutation_rate}, but this simulation used mutation rate {contig.mutation_rate}. ..."

apragsdale replied:
Good idea. Updated now to include those rates

apragsdale commented:
One obvious thing still missing are tests to ensure that the model mutation rate matches the QC model mutation rate (which have not yet been included, and should be compiled and added by someone other than myself). Am I missing anything else that is obvious here (or non-obvious)?

apragsdale commented:
Added a test for mutation rate equality between model and its QC, and also added a test for generation time equality (I also slightly refactored the qc register so that it registers the stdpopsim model object instead of just the msprime Demography object, so that it's easier to check other attributes of the model).

The tests that fail now are because the QC models need mutation rates to be added.

jeromekelleher commented on stdpopsim/catalog/AraTha/demographic_models.py:
```diff
@@ -102,6 +102,8 @@ def _sma_1pop():
         )
     ]
 
+    mutation_rate = 7.1e-9
```
Maybe we should separate out these catalog changes from the code updates? These should be QC'd more closely and would require looking up the original papers etc, which is a lot to do at once.

Specifying a mutation rate shouldn't be necessary for the models, it can default to None so the basic change PR can go through.

apragsdale replied:
Done - I'll open a separate PR with these rates and their tests for equality.

jeromekelleher commented on stdpopsim/cli.py:
```diff
@@ -293,6 +293,8 @@ def write_citations(engine, model, contig, species):
     """
     cite_str = ["If you use this simulation in published work, please cite:"]
     for citation in get_citations(engine, model, contig, species):
+        if "mutation rate" in citation.reasons and model.mutation_rate is not None:
```
We have constants for this I think, don't need to use a hard coded string. citations.CiteReason.MUT_RATE I think?

jeromekelleher commented on stdpopsim/cli.py:
```diff
@@ -587,7 +590,11 @@ def write_simulation_summary(engine, model, contig, samples, seed=None):
     mean_recomb_rate = contig.recombination_map.mean_rate
-    mut_rate = contig.mutation_rate
+    # use the model mutation rate, if provided with the demographic model
+    if model.mutation_rate is not None:
```
probably good to add a logging.info() saying where the mutation rate is coming from

apragsdale replied:
I'm not sure if what I have now here is exactly what you mean... let me know if you had something else in mind.

andrewkern replied:
could be good to print the actual rate in that logging message so the user can see it

jeromekelleher commented on stdpopsim/engines.py:
```diff
@@ -112,6 +112,20 @@ def _warn_zigzag(self, demographic_model):
+    def _warn_mutation_rate_mismatch(self, contig, demographic_model):
+        if (
+            demographic_model.mutation_rate is not None
+            and demographic_model.mutation_rate != contig.mutation_rate
```
math.isclose() would be better here, exact equality will lead to noise I think.

jeromekelleher commented on stdpopsim/models.py:
```diff
@@ -79,20 +82,22 @@ def __init__(
         self.generation_time = 1 if generation_time is None else generation_time
+        self.mutation_rate = mutation_rate
         self.citations = [] if citations is None else citations
-        self.qc_model = qc_model
```
What's the reasoning for this change?

apragsdale replied:
Previously, the QC model (just the `Demography` object) was attached to a model to check 1) that a model has been QC'd (is not None), and 2) to check for equivalency in tests. I think it would also be good to check equivalency of other attributes of the model to its QC counterpart, such as generation time and mutation rate, so that those are more easily available to compare to each other if the full QC stdpopsim DemographicModel object is attached instead.

apragsdale replied:
Previously, the QC model (just the Demography object) was attached to a model to check 1) that a model has been QC'd (is not None), and 2) to check for equivalency in tests. I think it would also be good to check equivalency of other attributes of the model to its QC counterpart, such as generation time and mutation rate, so that those are more easily available to compare to each other if the full QC stdpopsim DemographicModel object is attached instead.

petrelharp replied:
hm, but this is changing the API for the Demographic Model - if we really want to change this, we should probably leave `qc_model` as a deprecated synonym... but, this *is* a "model" after all, so `qc_model` doesn't seem so bad?

apragsdale replied:
That's true - I'm fine with either name, and that avoids a param name change. In the equivalency tests we would then check `DemographicModel.model == DemographicModel.qc_model.model`

andrewkern replied:
yeah i agree with Peter here. I think we should probably avoid changing the param name

jeromekelleher commented on stdpopsim/species.py:
```diff
@@ -169,6 +170,8 @@ def get_contig(
             same chromosome-specific mutation and recombination rates.
             This option cannot currently be used in conjunction with the
             ``genetic_map`` argument.
+        :param float mutation_rate: The per-base mutation rate. If none is given,
```
Why do we need a species mutation rate as well?

apragsdale replied:
Previously, the QC model (just the Demography object) was attached to a model to check 1) that a model has been QC'd (is not None), and 2) to check for equivalency in tests. I think it would also be good to check equivalency of other attributes of the model to its QC counterpart, such as generation time and mutation rate, so that those are more easily available to compare to each other if the full QC stdpopsim DemographicModel object is attached instead.

jeromekelleher replied:
I think this comment is about the ``qc_model`` -> ``qc`` change above? (I don't see the connection with Species.mutation_rate?)

apragsdale replied:
Yeah sorry, replied to wrong review.. edited comment.

jeromekelleher replied:
Hmm - Github seems to have messed up the comments here - ignore

jeromekelleher replied:
Hmm, well, let's take the Species.mutation_rate out of this PR and get the other changes merged first. It's not clear why we'd have a Species.mutation_rate and not a Species.recombination_rate, etc, and why this wouldn't be derived from the per-chromosome values that we're trying to catalog.

apragsdale replied:
This is in the `get_contig` function, so that you can specify a contig with the demographic model's mutation rate instead of the chromosome rate. I don't think there is a Species.mutation_rate still?

jeromekelleher replied:
Ah, sorry, my bad.

jeromekelleher reviewed (commented):
Basically looks good @apragsdale, but there's an awful lot changing here at once. I think we should at least separate this into the basic changes to the object model and the changes to the catalog data. I think we're also increasingly muddling up simulation logic with the job of declaratively encoding information about species in the catalog. To me, there should be no concept of the semantics of running simulations in the catalog code, it should just be about declaring information that we know about species.

I'm also increasingly negative about the idea of a "generic" contig - it really seems like more trouble than it's worth.

apragsdale commented:
> I think we should at least separate this into the basic changes to the object model and the changes to the catalog data.

That makes sense - I had gone through and looked up all mutation rates from implemented demographic models already, so I'll move them to their own PR that can be QC'd separately.

> I think we're also increasingly muddling up simulation logic with the job of declaratively encoding information about species in the catalog. To me, there should be no concept of the semantics of running simulations in the catalog code, it should just be about declaring information that we know about species.

I'm not sure I follow, what is being added to the catalog that is being muddled here, aside from mutation rates to the demographic models? (in my view, the mutation rates are just as model-specific as the generation times, population sizes, and demographic event times)

> I'm also increasingly negative about the idea of a "generic" contig - it really seems like more trouble than it's worth.

I think it's an important basic feature. I can only speak for my own usage, but I far more often simulate a fixed length (say 1Mb with a rec rate that I specify) than simulate under a listed chromosome.

apragsdale commented:
I think this fix is just about ready to merge after a rebase/squash. Once it is, I can open a PR with the catalog mutation rate changes and fix up the tests using those specified mutation rates.

petrelharp commented on stdpopsim/engines.py:
```diff
@@ -112,6 +113,19 @@ def _warn_zigzag(self, demographic_model):
+                f"{demographic_model.mutation_rate}, but this simulation used the "
+                f"contig's mutation rate {contig.mutation_rate}. Diversity levels "
+                "may be different than expected for this species. For details see "
+                "documenation at "
```
"documen**t**ation"

apragsdale replied:
Fixed! Good spot

apragsdale commented:
Just rebased. Would love to get this merged soon!

grahamgower reviewed (approved):
LGTM!

jeromekelleher commented on docs/tutorial.rst:
```diff
@@ -282,7 +293,7 @@ using SLiM with a (very large!) scaling factor of 1000, we could run
 
 .. code-block:: console
 
-   $ stdpopsim -e slim --slim-scaling-factor 1000 DroMel \
```
Does this work? I thought ``--engine`` options were to the top level parser and not available to the subcommands.

apragsdale replied:
Ah, you're right. It guess it looked weird to have the species "positional" argument coming in the middle of some optional flags. In general, the CLI interface is confusing and doesn't follow standard guidelines, but that's another matter. I'll change it back so it works with what we've got.

jeromekelleher commented on tests/test_models.py:
```diff
@@ -315,3 +327,34 @@ def test_zigzag_produces_warning(self):
+class TestMutationRates(unittest.TestCase):
+    def test_mutation_rate_warning(self):
+        species = stdpopsim.get_species("HomSap")
+        model = species.get_demographic_model("OutOfAfrica_3G09")
```
Better do a copy.deepcopy() on this object, as you're potentially modifying the global state and this might affect other test. Ditto below.

jeromekelleher reviewed (approved):
LGTM, minor comments.

apragsdale commented:
I have no clue how to test these logging outputs that codecov is complaining about...

agladstein commented:
Hey, I'm not sure if I'll be able to review this today. If you want to merge it go for it. I can always add an issue later if something comes up.

jeromekelleher commented:
I'm happy to merge this as it is, the coverage things look pretty minor.

apragsdale commented:
Is this ok to merge? I have a follow-up PR with updates to the catalog to include the mutation rate, but waiting for this to be pulled in.
